torch/csrc/distributed/c10d/ProcessGroup.hpp

```diff
@@ -105,173 +111,308 @@ class TORCH_API ProcessGroup : public torch::CustomClassHolder {
   virtual void endCoalescing(
       c10::DeviceType deviceType,
       std::vector<c10::intrusive_ptr<Work>>& reqs) {
-    // only nccl has implemented startCoalescing so only execute for nccl backends
+    // only nccl has implemented startCoalescing so only execute for nccl
+    // backends
     if (getBackendType() == BackendType::NCCL) {
       getBackend(deviceType)->endCoalescing(reqs);
     }
   }
 
-  // Consider using ops in Ops.hpp instead of the below, which route things
-  // to the dispatcher.
-  // TODO: Find a way to force the above rule programmatically.
   virtual c10::intrusive_ptr<Work> broadcast(
```

Comment:
Question: do we need to think about whether these collective APIs are still virtual or not?
i.e. is there any class that still inherits from `ProcessGroup`?